library/L1_Peripheral/lpc17xx/system_controller.hpp

```diff
@@ -117,273 +53,563 @@ class SystemController final : public sjsu::SystemController
-  /// on the input frequency provided to the PLL.
-  ///
-  /// @param frequency Frequency of the external oscillator used to drive PLL1.
-  void SetUsbPllInputFrequency(UsbPllInputFrequency frequency)
-  {
-    // NOTE:  It is very important not to merge any steps below. For example, do
-    //        not update the PLL1CFG and enable PLL1 simultaneously with the
-    //        same feed sequence.
 
-    // 1. Disconnect PLL1 with one feed sequence if PLL1 is already connected.
-    system_controller->PLL1CON =
-        bit::Clear(system_controller->PLL1CON, Pll::kConnectBit);
-    WritePllFeedSequence(PllSelect::kUsbPll);
-    // 2. Disable PLL1 with one feed sequence.
-    system_controller->PLL1CON =
-        bit::Clear(system_controller->PLL1CON, Pll::kEnableBit);
-    WritePllFeedSequence(PllSelect::kUsbPll);
-    // 3. Write to the PLL1CFG and make it effective with one feed sequence.
-    constexpr uint8_t kDivider = Value(UsbPllDivider::kDivideBy1);
-    // depending on the input frequency, the multiplier value will always be
-    // either 2, 3, or 4
-    const uint8_t kMultiplier[] = {
-      Value(UsbPllMultiplier::kMultiplyBy4),
-      Value(UsbPllMultiplier::kMultiplyBy3),
-      Value(UsbPllMultiplier::kMultiplyBy2),
-    };
-    system_controller->PLL1CFG =
-        bit::Insert(system_controller->PLL1CFG, kMultiplier[Value(frequency)],
-                    UsbPll::kMultiplier);
-    system_controller->PLL1CFG =
-        bit::Insert(system_controller->PLL1CFG, kDivider, UsbPll::kDivider);
-    WritePllFeedSequence(PllSelect::kUsbPll);
-    // 4. Enable PLL1 with one feed sequence.
-    system_controller->PLL1CON =
-        bit::Set(system_controller->PLL1CON, Pll::kEnableBit);
-    WritePllFeedSequence(PllSelect::kUsbPll);
-    // 5. Configurations to the PLL must be locked before it can be connected.
-    SJ2_ASSERT_FATAL(WaitForPllLockStatus(PllSelect::kUsbPll),
-                     "PLL1 lock could not be established before timeout");
-    // 6. Connect PLL1 with one feed sequence.
-    system_controller->PLL1CON =
-        bit::Set(system_controller->PLL1CON, Pll::kConnectBit);
-    WritePllFeedSequence(PllSelect::kUsbPll);
+  void PowerDownPeripheral(PeripheralID peripheral_select) const override
+  {
+    system_controller->PCONP =
+        bit::Clear(system_controller->PCONP, peripheral_select.device_id);
+  }
 
-    SJ2_ASSERT_FATAL(WaitForPllConnectionStatus(PllSelect::kUsbPll),
-                     "Failed to connect PLL1.");
+  units::frequency::hertz_t GetClockRate(PeripheralID peripheral) const override
+  {
+    switch (peripheral.device_id)
+    {
+      case Clocks::kCpu.device_id:
+      {
+        return cpu_clock_rate_;
+      }
+      case Clocks::kUsb.device_id:
+      {
+        return usb_clock_rate_;
+      }
+      default:  // All other peripherals
+      {
+        return cpu_clock_rate_ / GetPeripheralClockDivider(peripheral);
```

Comment:
```suggestion
        return cpu_clock_rate_;
```
See comment below for my reason for context for why I believe this should be this way.